runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -1330,6 +1338,19 @@ public void setTimer(
     @Deprecated
     @Override
     public void setTimer(TimerData timer) {
+      if (timer.getTimestamp().isAfter(GlobalWindow.INSTANCE.maxTimestamp())) {
```

Comment:
Could we do this using CleanupTimer? Something like
```java
      StatefulDoFnRunner.CleanupTimer cleanupTimer =
          new StatefulDoFnRunner.TimeInternalsCleanupTimer(timerInternals, windowingStrategy) {
            @Override
            public void setForWindow(Object input, BoundedWindow window) {
              if (!window.equals(GlobalWindow.INSTANCE)) {
                super.setForWindow(input, window);
              }
            }
          };
```
in DoFnOperator:317

Reply:
It was a deliberate decision not to do it using CleanupTimer. Setting the timer is an implementation detail and should be transparent to the CleanupTimer class.

Reply:
Also consider this requires to add the code at both implementations of CleanupTimer (portable and classic).

Reply:
The namespace we use for cleanup is based on `GlobalWindow` and so should be the condition here. I think we can prioritize clarity over the duplication of a single conditional return statement, why should be covered by the unit tests.

Reply:
We could create a NonGlobalWindowCleanupTimer (or something similar) in runner-core and reuse it in both portability and non-portable runner. If the code duplication would be the biggest concern.

Reply:
Please have a look, I've adjusted the implementation.